modules/solid_mechanics/src/vectorpostprocessors/CrackFrontNonlocalStress.C

```diff
@@ -8,158 +8,27 @@
-  params.addRequiredParam<UserObjectName>("crack_front_definition",
-                                          "The CrackFrontDefinition user object name");
-  params.addRequiredParam<Real>("box_length", "Distance in front of crack front.");
-  params.addRequiredParam<Real>("box_height", "Distance normal to front of crack front.");
-  params.addParam<Real>("box_width", "Distance tangent to front of crack front.");
-
-  params.addParam<std::string>("base_name",
-                               "Optional parameter that allows the user to define "
-                               "multiple mechanics material systems on the same "
-                               "block, i.e. for multiple phases");
-  params.set<bool>("use_displaced_mesh") = false;
-  // EXEC_NONLINEAR to work with xfem_udpates
-  params.set<ExecFlagEnum>("execute_on") = {EXEC_TIMESTEP_BEGIN};
+  InputParameters params = CrackFrontNonlocalMaterialBase::validParams();
   params.addClassDescription("Computes the average stress normal to the crack face.");
   return params;
 }
 
 CrackFrontNonlocalStress::CrackFrontNonlocalStress(const InputParameters & parameters)
-  : ElementVectorPostprocessor(parameters),
-    _box_length(getParam<Real>("box_length")),
-    _box_width(isParamValid("box_width") ? getParam<Real>("box_width") : 1),
-    _box_height(getParam<Real>("box_height")),
-    _base_name(isParamValid("base_name") ? getParam<std::string>("base_name") + "_" : ""),
-    _stress(getMaterialProperty<RankTwoTensor>(_base_name + "stress")),
-    _x(declareVector("x")),
-    _y(declareVector("y")),
-    _z(declareVector("z")),
-    _position(declareVector("id")),
-    _avg_crack_tip_stress(declareVector("crack_tip_stress"))
-{
-  if (_mesh.dimension() == 3 && !isParamSetByUser("box_width"))
-    paramError("box_width", "Must define box_width in 3D problems.");
-  // add user object dependencies by name (the UOs do not need to exist yet for this)
-  _depend_uo.insert(getParam<UserObjectName>("crack_front_definition"));
-}
-
-void
-CrackFrontNonlocalStress::initialSetup()
-{
-  // gather coupled user objects late to ensure they are constructed. Do not add them as
-  // dependencies (that's already done in the constructor).
-  const auto uo_name = getParam<UserObjectName>("crack_front_definition");
-  _crack_front_definition =
-      &(getUserObjectByName<CrackFrontDefinition>(uo_name, /*is_dependency = */ false));
-}
-
-void
-CrackFrontNonlocalStress::initialize()
-{
-  std::size_t num_pts = _crack_front_definition->getNumCrackFrontPoints();
-
-  _volume.assign(num_pts, 0.0);
-  _x.assign(num_pts, 0.0);
-  _y.assign(num_pts, 0.0);
-  _z.assign(num_pts, 0.0);
-  _position.assign(num_pts, 0.0);
-  _avg_crack_tip_stress.assign(num_pts, 0.0);
-}
-
-void
-CrackFrontNonlocalStress::execute()
+  : CrackFrontNonlocalMaterialBase(parameters),
+    _stress(getMaterialProperty<RankTwoTensor>(_base_name + getParam<std::string>("material_name")))
```

Comment:
Since this is specifically operating on stress, I prefer the original parameter ('stress') to 'material_name'. That would require moving the parameter from the base class into the derived classes, but I think that's worth doing.

Reply:
@dschwen helped me move this to the derived classes.